All tests should use string UIDs rather than integer UIDs

The following unit test:

 ```
@pytest.mark.parametrize("model_uid", rand_l(1, 500, 5))
def test_set_cube_uid_keeps_passed_uid_intact(
    mocker, default_setup, model_uid, comms, ui
):
```
Passes sometimes and fails sometimes.

It can be found in the script:
` medperf/cli/medperf/tests/commands/test_compatibility_test.py`

Most probably it fails for certain parameters that are randomly generated. To reproduce, replace ` rand_l(1, 500, 5)` in the parameterization with the list `[20,47,45]`. The following errors will be raised:

```
FAILED test_compatibility_test.py::test_set_cube_uid_keeps_passed_uid_intact[20] - AssertionError: assert 'test_1651680003' == 20
FAILED test_compatibility_test.py::test_set_cube_uid_keeps_passed_uid_intact[47] - AssertionError: assert 'test_1651680003' == 47
FAILED test_compatibility_test.py::test_set_cube_uid_keeps_passed_uid_intact[45] - AssertionError: assert 'test_1651680004' == 45
```

--- a/cli/medperf/commands/compatibility_test.py
+++ b/cli/medperf/commands/compatibility_test.py
@@ -116,8 +116,13 @@ class CompatibilityTestExecution:
             demo_dataset_url = None
             demo_dataset_hash = None
 
-        self.benchmark = Benchmark.tmp(self.data_prep, self.model, self.evaluator,
-                                            demo_dataset_url, demo_dataset_hash)
+        self.benchmark = Benchmark.tmp(
+            self.data_prep,
+            self.model,
+            self.evaluator,
+            demo_dataset_url,
+            demo_dataset_hash,
+        )
         self.benchmark_uid = self.benchmark.uid
 
     def execute_benchmark(self):

--- a/cli/medperf/commands/compatibility_test.py
+++ b/cli/medperf/commands/compatibility_test.py
@@ -103,16 +103,22 @@ class CompatibilityTestExecution:
         transformed to cube uids and benchmark is mocked/obtained.
         """
         if self.benchmark_uid:
-            self.benchmark = Benchmark.get(self.benchmark_uid, self.comms)
-            self.set_cube_uid("data_prep", self.benchmark.data_preparation)
-            self.set_cube_uid("model", self.benchmark.reference_model)
-            self.set_cube_uid("evaluator", self.benchmark.evaluator)
+            benchmark = Benchmark.get(self.benchmark_uid, self.comms)
+            self.set_cube_uid("data_prep", benchmark.data_preparation)
+            self.set_cube_uid("model", benchmark.reference_model)
+            self.set_cube_uid("evaluator", benchmark.evaluator)
+            demo_dataset_url = benchmark.demo_dataset_url
+            demo_dataset_hash = benchmark.demo_dataset_hash
         else:
             self.set_cube_uid("data_prep")
             self.set_cube_uid("model")
             self.set_cube_uid("evaluator")
-            self.benchmark = Benchmark.tmp(self.data_prep, self.model, self.evaluator)
-            self.benchmark_uid = self.benchmark.uid
+            demo_dataset_url = None
+            demo_dataset_hash = None
+
+        self.benchmark = Benchmark.tmp(self.data_prep, self.model, self.evaluator,
+                                            demo_dataset_url, demo_dataset_hash)
+        self.benchmark_uid = self.benchmark.uid
 
     def execute_benchmark(self):
         """Runs the benchmark execution flow given the specified testing parameters
@@ -136,7 +142,7 @@ class CompatibilityTestExecution:
             attr (str): Attribute to check and/or reassign.
             fallback (any): Value to assign if attribute is empty. Defaults to None.
         """
-        logging.info("Establishing model_uid for test execution")
+        logging.info(f"Establishing {attr}_uid for test execution")
         val = getattr(self, attr)
         if val is None:
             logging.info(f"Empty attribute: {attr}. Assigning fallback: {fallback}")
@@ -147,11 +153,12 @@ class CompatibilityTestExecution:
         val = str(val)
         if os.path.exists(val):
             logging.info("local path provided. Creating symbolic link")
-            self.model = config.test_cube_prefix + str(int(time()))
+            temp_uid = config.test_cube_prefix + str(int(time()))
+            setattr(self, attr, temp_uid)
             cubes_storage = storage_path(config.cubes_storage)
-            dst = os.path.join(cubes_storage, self.model)
+            dst = os.path.join(cubes_storage, temp_uid)
             os.symlink(val, dst)
-            logging.info(f"local cube will linked to path: {dst}")
+            logging.info(f"local cube will be linked to path: {dst}")
 
     def set_data_uid(self):
         """Assigns the data_uid used for testing according to the initialization parameters.

--- a/cli/medperf/commands/compatibility_test.py
+++ b/cli/medperf/commands/compatibility_test.py
@@ -132,6 +132,7 @@ class CompatibilityTestExecution:
             return
 
         # Check if value is a server UID
+        val = str(val)
         if os.path.exists(val):
             logging.info("local path provided. Creating symbolic link")
             self.model = config.test_cube_prefix + str(int(time()))

--- a/cli/medperf/commands/compatibility_test.py
+++ b/cli/medperf/commands/compatibility_test.py
@@ -78,11 +78,23 @@ class CompatibilityTestExecution:
         Specifically, a benchmark must be passed if any other workflow
         parameter is not passed.
         """
-        params = [self.data_uid, self.data_prep, self.model, self.evaluator]
+        params = [self.data_uid, self.model, self.evaluator]
         none_params = [param is None for param in params]
         if self.benchmark_uid is None and any(none_params):
             pretty_error(
                 "Invalid combination of arguments to test. Ensure you pass a benchmark or a complete mlcube flow",
+                self.ui,
+            )
+        # a redundant data preparation cube
+        if self.data_uid is not None and self.data_prep is not None:
+            pretty_error(
+                "Invalid combination of arguments to test. The passed preparation cube will not be used",
+                self.ui,
+            )
+        # a redundant benchmark
+        if self.benchmark_uid is not None and not any(none_params):
+            pretty_error(
+                "Invalid combination of arguments to test. The passed benchmark will not be used",
                 self.ui,
             )
 

--- a/cli/medperf/tests/commands/test_compatibility_test.py
+++ b/cli/medperf/tests/commands/test_compatibility_test.py
@@ -94,8 +94,9 @@ def test_prepare_test_gets_benchmark_or_tmp(mocker, uid, benchmark, comms, ui):
     # Assert
     if uid:
         get_spy.assert_called_once_with(uid, comms)
-    tmp_spy.assert_called_once_with(prep, model, eval,
-                                        bmk.demo_dataset_url, bmk.demo_dataset_hash)
+    tmp_spy.assert_called_once_with(
+        prep, model, eval, bmk.demo_dataset_url, bmk.demo_dataset_hash
+    )
 
 
 @pytest.mark.parametrize("uid", [None, "1"])
@@ -368,6 +369,7 @@ def test_download_demo_data_extracts_expected_paths(
     assert data_path == exp_data_path
     assert labels_path == exp_labels_path
 
+
 @pytest.mark.parametrize("data_uid", rand_l(1, 500, 1) + [None])
 @pytest.mark.parametrize("prep_uid", rand_l(1, 500, 1) + [None])
 @pytest.mark.parametrize("model_uid", rand_l(1, 500, 1) + [None])
@@ -392,26 +394,38 @@ def test_run_uses_correct_uids(
     mocker.patch(PATCH_TEST.format("Benchmark.get"), return_value=bmk)
     mocker.patch("os.path.exists", return_value=False)
     mocker.patch(PATCH_TEST.format("get_uids"), return_value=[str(data_uid)])
-    mocker.patch(PATCH_TEST.format("CompatibilityTestExecution.download_demo_data"),
-                                                                return_value=("", ""))
-    mocker.patch(PATCH_TEST.format("DataPreparation.run"), return_value=demo_dataset_uid)
+    mocker.patch(
+        PATCH_TEST.format("CompatibilityTestExecution.download_demo_data"),
+        return_value=("", ""),
+    )
+    mocker.patch(
+        PATCH_TEST.format("DataPreparation.run"), return_value=demo_dataset_uid
+    )
     mocker.patch(PATCH_TEST.format("Dataset"), return_value=dataset)
     mocker.patch(PATCH_TEST.format("Result"))
 
     tmp_spy = mocker.patch(PATCH_TEST.format("Benchmark.tmp"), return_value=tmp_bmk)
     exec_spy = mocker.patch(PATCH_TEST.format("BenchmarkExecution.run"))
-    
+
     exp_data_uid = demo_dataset_uid if data_uid is None else data_uid
     exp_prep_uid = bmk_prep_uid if prep_uid is None else prep_uid
     exp_model_uid = bmk_model_uid if model_uid is None else model_uid
     exp_eval_uid = bmk_eval_uid if eval_uid is None else eval_uid
 
     # Act
-    CompatibilityTestExecution.run(bmk_uid, comms, ui,
-                            data_uid, prep_uid, model_uid, eval_uid)
+    CompatibilityTestExecution.run(
+        bmk_uid, comms, ui, data_uid, prep_uid, model_uid, eval_uid
+    )
 
     # Assert
-    tmp_spy.assert_called_once_with(exp_prep_uid, exp_model_uid, exp_eval_uid,
-                                            bmk.demo_dataset_url, bmk.demo_dataset_hash)
-    exec_spy.assert_called_once_with(tmp_uid, exp_data_uid, exp_model_uid,
-                                        comms, ui, run_test=True)
+    tmp_spy.assert_called_once_with(
+        exp_prep_uid,
+        exp_model_uid,
+        exp_eval_uid,
+        bmk.demo_dataset_url,
+        bmk.demo_dataset_hash,
+    )
+    exec_spy.assert_called_once_with(
+        tmp_uid, exp_data_uid, exp_model_uid, comms, ui, run_test=True
+    )
+

--- a/cli/medperf/tests/commands/test_compatibility_test.py
+++ b/cli/medperf/tests/commands/test_compatibility_test.py
@@ -428,4 +428,3 @@ def test_run_uses_correct_uids(
     exec_spy.assert_called_once_with(
         tmp_uid, exp_data_uid, exp_model_uid, comms, ui, run_test=True
     )
-

--- a/cli/medperf/tests/commands/test_compatibility_test.py
+++ b/cli/medperf/tests/commands/test_compatibility_test.py
@@ -18,6 +18,8 @@ def benchmark(mocker):
         bmk.data_preparation = data_prep
         bmk.reference_model = reference_model
         bmk.evaluator = evaluator
+        bmk.demo_dataset_url = None
+        bmk.demo_dataset_hash = None
         return bmk
 
     return benchmark_gen
@@ -34,6 +36,7 @@ def dataset(mocker):
 def default_setup(mocker, benchmark, dataset):
     bmk = benchmark(1, 1, 2, 3)
     mocker.patch(PATCH_TEST.format("Benchmark.get"), return_value=bmk)
+    mocker.patch(PATCH_TEST.format("Benchmark.tmp"), return_value=bmk)
     mocker.patch(
         PATCH_TEST.format("CompatibilityTestExecution.download_demo_data"),
         return_value=("", ""),
@@ -46,9 +49,10 @@ def default_setup(mocker, benchmark, dataset):
 @pytest.mark.parametrize(
     "test_params",
     [
-        ((None, "data", "prep", "model", "eval"), True),
+        ((None, "data", "prep", "model", "eval"), False),
+        (("bmk", "data", None, "model", "eval"), False),
+        ((None, "data", None, "model", "eval"), True),
         ((None, None, "prep", "model", "eval"), False),
-        ((None, "data", None, "model", "eval"), False),
         ((None, "data", "prep", None, "eval"), False),
         ((None, "data", "prep", "model", None), False),
     ],
@@ -69,7 +73,7 @@ def test_validate_fails_if_incomplete_tmp_benchmark_passed(
     if should_be_valid:
         spy.assert_not_called()
     else:
-        spy.assert_called_once()
+        spy.assert_called()
 
 
 @pytest.mark.parametrize("uid", [None, "1"])
@@ -90,8 +94,8 @@ def test_prepare_test_gets_benchmark_or_tmp(mocker, uid, benchmark, comms, ui):
     # Assert
     if uid:
         get_spy.assert_called_once_with(uid, comms)
-    else:
-        tmp_spy.assert_called_once_with(prep, model, eval)
+    tmp_spy.assert_called_once_with(prep, model, eval,
+                                        bmk.demo_dataset_url, bmk.demo_dataset_hash)
 
 
 @pytest.mark.parametrize("uid", [None, "1"])
@@ -279,6 +283,7 @@ def test_run_returns_uids(
     # Arrange
     bmk = benchmark(bmk_uid, data_uid, model_uid, "3")
     mocker.patch(PATCH_TEST.format("Benchmark.get"), return_value=bmk)
+    mocker.patch(PATCH_TEST.format("Benchmark.tmp"), return_value=bmk)
     mocker.patch(PATCH_TEST.format("CompatibilityTestExecution.validate"))
     mocker.patch(PATCH_TEST.format("CompatibilityTestExecution.set_cube_uid"))
     mocker.patch(PATCH_TEST.format("CompatibilityTestExecution.set_data_uid"))
@@ -286,7 +291,6 @@ def test_run_returns_uids(
         PATCH_TEST.format("CompatibilityTestExecution.execute_benchmark"),
         return_value=results,
     )
-    mocker.patch(PATCH_TEST.format("Benchmark.write"))
 
     # Act
     ret_uids = CompatibilityTestExecution.run(
@@ -309,6 +313,7 @@ def test_download_demo_data_fails_if_incorrect_hash(mocker, benchmark, comms, ha
     bmk.demo_dataset_url = "url"
     bmk.demo_dataset_hash = hash
     mocker.patch(PATCH_TEST.format("Benchmark.get"), return_value=bmk)
+    mocker.patch(PATCH_TEST.format("Benchmark.tmp"), return_value=bmk)
     mocker.patch.object(comms, "get_benchmark_demo_dataset", return_value=("", ""))
     mocker.patch(PATCH_TEST.format("get_file_sha1"), return_value="hash")
     exec = CompatibilityTestExecution(uid, data, prep, model, eval, comms, ui)
@@ -341,6 +346,7 @@ def test_download_demo_data_extracts_expected_paths(
     bmk.demo_dataset_url = "url"
     bmk.demo_dataset_hash = "hash"
     mocker.patch(PATCH_TEST.format("Benchmark.get"), return_value=bmk)
+    mocker.patch(PATCH_TEST.format("Benchmark.tmp"), return_value=bmk)
     mocker.patch.object(comms, "get_benchmark_demo_dataset", return_value=("", ""))
     mocker.patch(PATCH_TEST.format("get_file_sha1"), return_value="hash")
 
@@ -361,3 +367,51 @@ def test_download_demo_data_extracts_expected_paths(
     # Assert
     assert data_path == exp_data_path
     assert labels_path == exp_labels_path
+
+@pytest.mark.parametrize("data_uid", rand_l(1, 500, 1) + [None])
+@pytest.mark.parametrize("prep_uid", rand_l(1, 500, 1) + [None])
+@pytest.mark.parametrize("model_uid", rand_l(1, 500, 1) + [None])
+@pytest.mark.parametrize("eval_uid", rand_l(1, 500, 1) + [None])
+def test_run_uses_correct_uids(
+    mocker, benchmark, dataset, data_uid, prep_uid, model_uid, eval_uid, comms, ui
+):
+    # Arrange
+    bmk_uid = "1"
+    bmk_prep_uid = "b1"
+    bmk_model_uid = "b2"
+    bmk_eval_uid = "b3"
+    demo_dataset_uid = "d1"
+    tmp_uid = "t1"
+
+    tmp_bmk = benchmark(tmp_uid, bmk_prep_uid, bmk_model_uid, bmk_eval_uid)
+    bmk = benchmark(bmk_uid, bmk_prep_uid, bmk_model_uid, bmk_eval_uid)
+    bmk.demo_dataset_url = "url"
+    bmk.demo_dataset_hash = "hash"
+
+    mocker.patch(PATCH_TEST.format("CompatibilityTestExecution.validate"))
+    mocker.patch(PATCH_TEST.format("Benchmark.get"), return_value=bmk)
+    mocker.patch("os.path.exists", return_value=False)
+    mocker.patch(PATCH_TEST.format("get_uids"), return_value=[str(data_uid)])
+    mocker.patch(PATCH_TEST.format("CompatibilityTestExecution.download_demo_data"),
+                                                                return_value=("", ""))
+    mocker.patch(PATCH_TEST.format("DataPreparation.run"), return_value=demo_dataset_uid)
+    mocker.patch(PATCH_TEST.format("Dataset"), return_value=dataset)
+    mocker.patch(PATCH_TEST.format("Result"))
+
+    tmp_spy = mocker.patch(PATCH_TEST.format("Benchmark.tmp"), return_value=tmp_bmk)
+    exec_spy = mocker.patch(PATCH_TEST.format("BenchmarkExecution.run"))
+    
+    exp_data_uid = demo_dataset_uid if data_uid is None else data_uid
+    exp_prep_uid = bmk_prep_uid if prep_uid is None else prep_uid
+    exp_model_uid = bmk_model_uid if model_uid is None else model_uid
+    exp_eval_uid = bmk_eval_uid if eval_uid is None else eval_uid
+
+    # Act
+    CompatibilityTestExecution.run(bmk_uid, comms, ui,
+                            data_uid, prep_uid, model_uid, eval_uid)
+
+    # Assert
+    tmp_spy.assert_called_once_with(exp_prep_uid, exp_model_uid, exp_eval_uid,
+                                            bmk.demo_dataset_url, bmk.demo_dataset_hash)
+    exec_spy.assert_called_once_with(tmp_uid, exp_data_uid, exp_model_uid,
+                                        comms, ui, run_test=True)

